test/test_flop_counter.py

```diff
@@ -0,0 +1,124 @@
+
+    def test_op(self):
+        mode = FlopCounterMode()
+        with mode:
+            torch.mm(T(4, 5), T(5, 6))
+        # 4 * 6 * 2 * 5 = 240
+        self.assertExpectedInline(get_total_flops(mode), """240""")
+
+        with mode:
+            torch.bmm(T(3, 4, 5), T(3, 5, 6))
+        # 3 * 4 * 6 * 2 * 5 = 720
+        self.assertExpectedInline(get_total_flops(mode), """720""")
+
+        with mode:
+            torch.addmm(T(4, 6), T(4, 5), T(5, 6))
+            torch.addmm(T(4, 1), T(4, 5), T(5, 6))
+            torch.addmm(T(6), T(4, 5), T(5, 6))
+
+        # 4 * 6 * 2 * 5 + 4 * 6 = 264
+        self.assertExpectedInline(get_total_flops(mode), """792""")
+
+        with mode:
+            torch.baddbmm(T(3, 4, 6), T(3, 4, 5), T(3, 5, 6))
+
+        # 3 * 4 * 6 * 2 * 5 + 3 * 4 * 6 = 792
+        self.assertExpectedInline(get_total_flops(mode), """792""")
+
+        with mode:
+            torch.conv2d(T(2, 3, 6, 6), T(6, 3, 4, 4), padding=1)
+
+        # out_image_size = 2 * 5 * 5
+        # kernel_size = 4 * 4
+        # c_out = 6
+        # c_in = 3
+        # out_image_size * kernel_size * c_out * 2 * c_in
+
+        # NB: I don't think this properly accounts for padding?
+        self.assertExpectedInline(get_total_flops(mode), """28800""")
+
+        with mode:
+            torch.conv1d(T(2, 3, 6), T(6, 3, 4), padding=1)
+
+        # out_image_size = 2 * 5
+        # kernel_size = 4
+        # c_out = 6
+        # c_in = 3
+        # out_image_size * kernel_size * c_out * 2 * c_in
+
+        # NB: I don't think this properly accounts for padding?
+        self.assertExpectedInline(get_total_flops(mode), """1440""")
+
+    def test_backward(self):
+        mode = FlopCounterMode()
+        with mode:
+            a = T(4, 5)
+            a = a.requires_grad_(True)
+            a = torch.mm(a, T(5, 6))
+            a = a.unsqueeze(0).expand(7, 4, 6)
+            a = torch.bmm(a, T(7, 6, 7))
+            a.sum().backward()
+
+        self.assertExpectedInline(get_total_flops(mode), """5184""")
+
+    @skipIfNoTorchVision
+    def test_module(self):
+        resnet18 = torchvision_models.resnet18()
+        mode = FlopCounterMode(resnet18)
+        with mode:
+            a = T(1, 3, 224, 224)
+            resnet18(a)
+
+        self.assertExpectedInline(get_total_flops(mode), """3628147688""")
+        layer1_conv_flops = mode.flop_counts['layer1'][torch.ops.aten.convolution]
+        layer1_conv_back_flops = mode.flop_counts['layer1'][torch.ops.aten.convolution_backward]
+        self.assertExpectedInline(str(layer1_conv_flops), """924844032""")
+        self.assertExpectedInline(str(layer1_conv_back_flops), """0""")
+
+    def test_custom(self):
+        mode = FlopCounterMode(custom_mapping={torch.ops.aten.add: lambda *args: 5})
+        with mode:
+            a = T(4, 5)
+            a + a
+
+        self.assertExpectedInline(get_total_flops(mode), """5""")
+
+
```

Comment:
Other cases to test:
* Module A call a submodule `self.subm.method_name()`, instead of `self.subm.__call__`.  It's ambiguous in this case which module (self or self.subm) the flops should attribute to, but please test that the results are at least reasonable.
* Module with scripted submodule: https://github.com/facebookresearch/fvcore/blob/51092b5515cbb493f73de079743dd6b11cc4bbf1/tests/test_flop_count.py#L832-L845
* Module with legacy autograd function: https://github.com/facebookresearch/fvcore/blob/51092b5515cbb493f73de079743dd6b11cc4bbf1/tests/test_flop_count.py#L820-L830  Not sure if this could be supported at all.

Reply:
> Module with scripted submodule:

Unfortunately we can't support hierarchies on scripted modules, since they don't support hooks. I'll add a test for just capturing flops under scripted functions.

> Module with legacy autograd function:

This works totally fine.

Reply:
Actually, today I just found that autograd function isn't working in the way I expected.

Usually users will put custom cuda kernels, which pytorch knows nothing about, in such autograd function. For example like this https://github.com/facebookresearch/detectron2/blob/e9f7e2ba15abd7badcb05ef6f5076f06b36a9c5b/detectron2/layers/deform_conv.py#L59

It seems in this PR there is no way to register a custom flop counter for custom kernels that are not registered as a pytorch op (e.g. https://github.com/facebookresearch/detectron2/blob/e9f7e2ba15abd7badcb05ef6f5076f06b36a9c5b/detectron2/layers/csrc/vision.cpp#L83-L94).

However, the fvcore flop counter can interpret the autograd function as an atomic unit (instead of entering it, like this PR), so it actually allows users to count custom kernel properly, for example in this test: https://github.com/facebookresearch/fvcore/blob/e8d19df2ffdaf0a78c8d88d6a2522c36b0cacb07/tests/test_flop_count.py#L820-L830